language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToLocalVarCodeAction.java

```diff
@@ -0,0 +1,136 @@
+        return CodeActionNodeValidator.validate(context.nodeAtRange());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<CodeAction> getCodeActions(CodeActionContext context,
+                                           RangeBasedPositionDetails posDetails) {
+        
+        if (context.currentSyntaxTree().isEmpty() || context.currentSemanticModel().isEmpty()) {
+            return Collections.emptyList();
+        }
+        
+        Node node = posDetails.matchedCodeActionNode(); 
+        if ((node.kind() == SyntaxKind.MAPPING_CONSTRUCTOR && node.parent().kind() == SyntaxKind.TABLE_CONSTRUCTOR) 
+                || (node.kind() == SyntaxKind.FUNCTION_CALL && node.parent().kind() == SyntaxKind.LOCAL_VAR_DECL)) {
+            return Collections.emptyList();
+        }
+        
+        
+        String varName = getLocalVarName(context);
+        String value = node.toSourceCode().strip();
+        LineRange replaceRange = node.lineRange();
+        Optional<TypeSymbol> typeSymbol = context.currentSemanticModel().get().typeOf(node);
+        if (typeSymbol.isEmpty() || typeSymbol.get().typeKind() == TypeDescKind.COMPILATION_ERROR) {
+            return Collections.emptyList();
+        }
+
+        Node statementNode = node;
+        while (statementNode != null && !(statementNode instanceof StatementNode) 
+                && !(statementNode instanceof ModuleMemberDeclarationNode)) {
+            statementNode = statementNode.parent();
+        }
+        if (statementNode == null) {
+            return Collections.emptyList();
+        }
+        String offset = " ".repeat(statementNode.lineRange().startLine().offset());
+        String varDeclStr = String.format("%s %s = %s;%n%s", typeSymbol.get().signature(), varName, value, offset);
+        Position varDeclPos = new Position(statementNode.lineRange().startLine().line(), 
+                statementNode.lineRange().startLine().offset());
+        TextEdit varDeclEdit = new TextEdit(new Range(varDeclPos, varDeclPos), varDeclStr);
+        TextEdit replaceEdit = new TextEdit(new Range(PositionUtil.toPosition(replaceRange.startLine()),
+                PositionUtil.toPosition(replaceRange.endLine())),  varName);
+
+        return Collections.singletonList(CodeActionUtil.createCodeAction(CommandConstants.EXTRACT_TO_VARIABLE, 
+                List.of(varDeclEdit, replaceEdit), context.fileUri(), CodeActionKind.RefactorExtract));
+    }
+
+    private String getLocalVarName(CodeActionContext context) {
+        Position pos = context.range().getEnd();
+        Set<String> allNames = context.visibleSymbols(new Position(pos.getLine(), pos.getCharacter())).stream()
+                .map(Symbol::getName)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toSet());
+        
+        return NameUtil.generateTypeName(VARIABLE_NAME_PREFIX, allNames);
```

Comment:
Shall we change the logic to suggest localVarName `variable2` after `variable1` instead of `variable12`?

Reply:
Fixed.